Tests/SwinjectTests/ContainerTests.swift

```diff
@@ -90,6 +90,69 @@ class ContainerTests: XCTestCase {
         XCTAssertNil(weakCat)
     }
 
+    func testShadowedRegistration_receiverHierarchyAccess() {
+        let parent = Container()
+        let child = Container(parent: parent, resolverHierarchyAccess: .receiver)
+
+        parent.register(Animal.self, factory: { _ in Dog()})
+        child.register(Animal.self, factory: { _ in Cat()})
+
+        parent.register(Animal.self, name: "Spot", factory: { resolver in
+            resolver.resolve(Animal.self)!
+        })
+
+        XCTAssert(child.resolve(Animal.self, name: "Spot") is Cat)
+        XCTAssert(parent.resolve(Animal.self, name: "Spot") is Dog)
+    }
+
+    func testShadowedRegistration_receiverHierarchyAccess_inObjectScopeContainer() {
+        let parent = Container()
+        let child = Container(parent: parent, resolverHierarchyAccess: .receiver)
+
+        parent.register(Animal.self, factory: { _ in Dog()})
+        child.register(Animal.self, factory: { _ in Cat()})
+
+        parent.register(Animal.self, name: "Spot", factory: { resolver in
+            resolver.resolve(Animal.self)!
+        })
+        .inObjectScope(.container)
+
+        XCTAssert(child.resolve(Animal.self, name: "Spot") is Cat)
+        // The instance provided by the child leaks into the parent
+        XCTAssert(parent.resolve(Animal.self, name: "Spot") is Cat)
```

Comment:
Also want to call attention to this test case as this behavior is the most unexpected/problematic of the way things currently work